### ISSUE TITLE
Error stringer.StringEncryptionTransformer

### ISSUE THREAD

**Issue by maxim19116** [java-deobfuscator/deobfuscator#90]
Deobfuscation failed. Please open a ticket on GitHub
java.lang.NullPointerException
        at java.lang.reflect.Array.get(Native Method)
        at com.javadeobfuscator.deobfuscator.executor.MethodExecutor.executeArrayLoad(MethodExecutor.java:101)
        at com.javadeobfuscator.deobfuscator.executor.MethodExecutor.execute(MethodExecutor.java:439)
        at com.javadeobfuscator.deobfuscator.executor.MethodExecutor.execute(MethodExecutor.java:80)
        at com.javadeobfuscator.deobfuscator.transformers.stringer.StringEncryptionTransformer.decrypt(StringEncryptionTransformer.java:204)
        at com.javadeobfuscator.deobfuscator.transformers.stringer.StringEncryptionTransformer.transform(StringEncryptionTransformer.java:61)
        at com.javadeobfuscator.deobfuscator.Deobfuscator.start(Deobfuscator.java:170)
        at com.javadeobfuscator.deobfuscator.DeobfuscatorMain.run(DeobfuscatorMain.java:102)
        at com.javadeobfuscator.deobfuscator.DeobfuscatorMain.main(DeobfuscatorMain.java:26)

**Comment by Eyremba** [comment]
The deobfuscator is outdated as fuck and no longer working. You will see such errors also when a jar is obfuscated using the latest version of Zelix or Allatori.

**Comment by FFY00** [comment]
> The deobfuscator is outdated as fuck and no longer working. You will see such errors also when a jar is obfuscated using the latest version of Zelix or Allatori.
By all means, submit a pull request to the project and help.

**Comment by Eyremba** [comment]
> By all means, submit a pull request to the project and help.

Why should I help....? I am pro Obfuscation and contra Deobfuscation, so I'm blissfully happy and overjoyed that all current Obfuscators can't be deobfuscated by tools like this. :)

**Comment by Janmm14** [comment]
Well there's a reason I do not publish my working deobfuscators too: People will find this project and report it to the obfuscation creators.

**Comment by FFY00** [comment]
> Well there's a reason I do not publish my working deobfuscator changes too: People will find this project and report it to the obfuscation creators. Or obfuscazion creators are looking here directly.
> 
> Samczsun is btw. working on a full controllable jre written in java as far as I know.

That's pretty cool, I will keep an eye out

> Why should I help....? I am pro Obfuscation and contra Deobfuscation, so I'm blissfully happy and overjoyed that all current Obfuscators can't be deobfuscated by tools like this. :)

Please close the page then, if you're not here to make constructive contributions stop wasting your time. And no, there are working deobfuscators but there's no point on releasing them as @Janmm14 said. I write my own deobfuscators as needed, java bytecode isn't that hard and bypassing obfuscators is just easy, it may take some time but still easy.

**Comment by Eyremba** [comment]
> Please close the page then, if you're not here to make constructive contributions stop wasting your time. And no, there are working deobfuscators but there's no point on releasing them as @Janmm14 said. I write my own deobfuscators as needed, java bytecode isn't that hard and bypassing obfuscators is just easy, it may take some time but still easy.

This statement clearly shows that you have no knowledge at all. Bypassing them is really hard, if the obfuscation is done wrong. I would bet that you have never seen real obfuscation. Nearly 90% of the Java kids out there use demo versions of obfuscators, or leaked versions which are outdated for more than 4 years now. Real obfuscation is fucking hard to deobfuscate. Some very good Java developers got some examples from me weeks and months ago, and they are still biting their teeths out while they try to deobfuscate them.

**Comment by FFY00** [comment]
> This statement clearly shows that you have no knowledge at all. Bypassing them is really hard, if the obfuscation is done right. I would bet that you have never seen real obfuscation. Nearly 90% of the Java kids out there use demo versions of obfuscators, or leaked versions which are outdated for more than 4 years now. Real obfuscation is fucking hard to deobfuscate. Some very good Java developers got some examples from me weeks and months ago, and they are still biting their teeths out while they try to deobfuscate them.

I wrote a metamorphic in assembly as my graduation thesis, why wouldn't I be able to bypass obfuscation? Java bytecode is just like a higher level assembly. I won't waste any more of my time talking to you, I know substantially more than you in this subject and what would be easy for me can be hard for you, I understand that and so should you. Have a nice day.

**Comment by Eyremba** [comment]
> I know substantially more than you in this subject

Um.... No. :) 

> what would be easy for me can be hard for you

Nope. It's hard for you too because if someone obfuscates a JAR file really really hard, you will give up because you have no chance.

**Comment by JoachimVandersmissen** [comment]
Hey @Eyremba, can I ask what your credentials here are? I mean, I like obfuscation as much as the next guy, and I don't really like deobfuscation tools either, but I mean, what have you done? Your github profile is 6 sources out of 171 repos, and 5 out of those 6 are minecraft plugins. I suppose you have a "string obfuscator" repository, but it does nothing except output some java source code. Is there anything you can show to back up your expertise? A linkedin profile maybe? At least make sure you know what you're talking about before you get into an argument.

**Comment by Eyremba** [comment]
@JoachimVandersmissen I know what I'm talking about, I also develop obfuscation tools, but of course private, if I would post them here on Github, it would be much more easy to deobfuscate their work.

**Comment by DarkSeraphim** [comment]
@Eyremba Even if you write the tools, it doesn't mean you know whether it can be cracked or not (or how hard it is). The reason is similar to why developers are not advised to roll out their own encryption algorithms: even if you think it's safe, you don't see all the issues concerning the algorithm.

Someone who's well versed in bytecode isn't stopped by obfuscation, since you can't obfuscate bytecode (unless you write native apps, but that's out of scope here), and reading bytecode is not a problem.

As long as obfuscation is a deterministic process (and yes, it pretty much always is in a sense, otherwise it's likely not reliable), it's not hard to crack.

(O, and "very good Java developers" is a bit of a meaningless term. 1) Do we really know / trust the skill of your testers, and 2) even if they're good Java developers, Java skills are not that important when analysing bytecode, since it's a lot lower level)

tl;dr; you're being way to narrow minded right now, the world is way bigger than you might guess :stuck_out_tongue:

**Comment by Eyremba** [comment]
> Someone who's well versed in bytecode isn't stopped by obfuscation, since you can't obfuscate bytecode

But if the Bytecode is 2000% times larger than before due to the obfuscation, it will be much harder and take much longer to understand everything. :)

**Comment by DarkSeraphim** [comment]
In most cases it's pretty repetitive, obfuscation generally requires you to find patterns, so you can replace them with it's smaller equivalent.

And as is known, practice makes perfect :P

**Comment by Eyremba** [comment]
> In most cases it's pretty repetitive, obfuscation generally requires you to find patterns, so you can replace them with it's smaller equivalent.

But if everything is flow-obfuscated in different ways, and everything is replaced by InvokeDynamics and by Reflection API calls, and again flow-ofbsucated in different ways, than you can not dispute that it will be a very time consuming task ;)

**Comment by DarkSeraphim** [comment]
Well you're free to provide a sample

**Comment by Eyremba** [comment]
> Well you're free to provide a sample

Just as a small sample, 2 class files from a jar file that has been obfuscated by my friend. You maybe know him too.

http://workupload.com/file/5KmCSTh

**Comment by DarkSeraphim** [comment]
Is this the encryption you mentioned before (which was hard to crack, and your friends didn't succeed)?

(I'm not near my laptop right now, so it'll have to wait a bit, but I'm asking already to date my curiosity :P)

**Comment by Eyremba** [comment]
> Is this the encryption you mentioned before (which was hard to crack, and your friends didn't succeed)?

No, but this are 2 classes from a very very good obfuscated jar file by my friend ;)

**Comment by konsolas** [comment]
@Eyremba it certainly is very good. In fact, it uses Zelix Klassmaster + 5 layers of stringer, the two industry leading obfuscators. It even leaves out the decryption classes so the obfuscation is literally unbreakable.

Also some dude added random while(false) loops and random DUP POP sequences, but those are optimized away instantly by krakatau. But that's not very useful.

**Comment by Eyremba** [comment]
>  In fact, it uses Zelix Klassmaster + 5 layers of stringer, the two industry leading obfuscators.

The obfuscators used in this two classes Are Zelix Klassmaster 8, Stringer 3.0.3, Allatori, and a custome one. However, it does not use "5 layers of Stringer", Stringer is only applied one time.

>  It even leaves out the decryption classes so the obfuscation is literally unbreakable.

Like I already said; 

> Just as a small sample, 2 class files from a jar file that has been obfuscated by my friend.

So it is not a whole jar, only 2 classes out of it, so yes it is possible that some things are missing because the heavy layers of obfuscation spread the decryption methods over all classes inside the jar file.

**Comment by DarkSeraphim** [comment]
From what I've heard, it appears to be pretty easy to deobfuscate xD (someone decompiled it for me, which apparently produced valid code)

**Comment by Eyremba** [comment]
> From what I've heard, it appears to be pretty easy to deobfuscate xD

You should prove it instead of just saying that.^^

Also, it is impossible to really deobfuscate this example because, like I said above, nearly all decryption classes are missing in this example here. :)

**Comment by DarkSeraphim** [comment]
@Eyremba well obviously, in a proper example you have a full jar file. Till
the extent of the code that we have, you can simply remove patterns like
while (false) {}, extract string decryptors and run them, etc.

On 16 Apr 2017 2:11 p.m., "Eyremba" <notifications@github.com> wrote:

From what I've heard, it appears to be pretty easy to deobfuscate xD

You should prove it instead of just saying that.^^

Also, it is impossible to really deobfuscate this example because, like I
said above, nearly all decryption classes are missing in this example here.
:)

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/java-deobfuscator/deobfuscator/issues/90#issuecomment-294348471>,
or mute the thread

**Comment by Eyremba** [comment]
> you can simply remove patterns like while (false) {}

But it is not "that simple" like you say it, because you have to do this thousands of times, and it can be hard to do this with automatic tools because they do not have the same patterns everytime depending on the flow-obfuscation.

> extract string decryptors and run them, etc.

But if you have many of them and ca not extract them easily? ;) Some obfuscators use so many variables, class names etc. in their encryption/decryption process, that you can not extract it so easily because it would not have all keys then and it could not run.

**Comment by samczsun** [comment]
Hey, let me just throw in my two cents here.

1) No one's asking you to post your obfuscator on GitHub. No one expects Stringer or Zelix or Allatori to go open source. However, if you plan on using your obfuscator at all then some samples are going to make their way to the general public. That's just logic. Unless you're worried someone'll crack it the moment they see it, why not release a sample when we ask?

2) Analysts don't really "give up" because someone barfed layers upon layers of shitty "obfuscation" onto a program. Why do you think even the more secure ransomeware are disassembled and analyzed eventually?

3) Sacrificing performance for security is not a good tradeoff. I could easily spam a class with a few thousand ldcs and crash basically any decompiler. It would also inflate runtime by 500% so it's not really justifiable.

4) The more layers of reflection/invokedynamic/flow obfuscation you add, the slower your program will run.  Meanwhile, all an ideal deobfuscator has to do is replace the reflection with direct calls, invokedynamic with Java 7 variants, one layer at a time.

5) You realize variable names, class names, etc are completely irrelevant to automated deobfuscation, right?

Finally, a few notes about your friend's "HardCoreObfuscation", since you can't even provide a sample of your own.

```L12305: dup 
L12306: iconst_1 
L12307: dup 
L12308: dup 
L12309: pop2 
L12310: dup 
L12311: dup 
L12312: pop2 
L12313: dup 
L12314: dup 
L12315: pop2 
L12316: dup ```

What even is this garbage. You're literally going to kill performance at the cost of something a simple peephole optimization could remove.

Furthermore, your friend should try and see how his obfuscator performs, without hiding behind Zelix and Stringer and the like. You don't see Allatori obfuscating their program with Stringer and bragging about how secure they are.

In fact, the only code I could find that doesn't look like industry standard obfuscation is this:

```
    private static Object an(Object a0, Object a1, Object a2)
    {
        java.lang.invoke.ConstantCallSite a3 = null;
        try
        {
            a3 = new java.lang.invoke.ConstantCallSite(((java.lang.invoke.MethodHandles$Lookup)a0).unreflect(me.Manic97.HardCoreObfuscation.q.h(Integer.valueOf((String)a1, 32).intValue())).asType((java.lang.invoke.MethodType)a2));
        }
        catch(ClassNotFoundException | IllegalAccessException a4)
        {
            throw new BootstrapMethodError((Throwable)a4);
        }
        return a3;
    }
```

Which is not even impressive. Just more invokedynamic obfuscation? Bleh.

**Comment by Eyremba** [comment]
> catch(ClassNotFoundException | IllegalAccessException a4)
        {
            throw new BootstrapMethodError((Throwable)a4);
        }
        return a3;

That is not real code, it is also a part of Stringers obfuscation^^

But yes I know what you mean. The problem is to find the perfect way between security and performance. Obfuscation can always be undone, I know this, but the time factor which is needed for this can be increased. But yes, it is important to find a way between having much security but a bad performance or having a good performance but no security.

**Comment by FFY00** [comment]
@maxim19116 Try again with the latest version.

**Comment by samczsun** [comment]
Please open a new issue if this still isn't resolved
